File: src/thingsync/things_source.py
"""Adapter: things.py's dicts in, :class:`ThingsTodo` out.

One of only two modules that touch the outside world. The readers are injected
so the whole transformation — including breadcrumb resolution — is testable
without a Things database.
"""

from __future__ import annotations

from collections.abc import Callable, Iterable, Mapping

from thingsync.model import ThingsProject, ThingsTodo


def _default_tasks(**kwargs):
    import things

    return things.tasks(**kwargs)


def _default_projects(**kwargs):
    import things

    return things.projects(**kwargs)


def load_projects(
    projects: Callable[..., Iterable[Mapping]] | None = None,
) -> list[ThingsProject]:
    """Every Things project, regardless of status.

    ``status`` defaults to ``'incomplete'`` in things.py, which would silently
    drop completed and cancelled projects — exactly the ones whose Reminders
    list needs to be torn down.
    """
    projects = projects or _default_projects
    return [
        ThingsProject(
            uuid=row["uuid"],
            title=row.get("title") or "",
            status=row.get("status") or "incomplete",
        )
        for row in projects(status=None)
    ]


def load_todos(
    tasks: Callable[..., Iterable[Mapping]] | None = None,
    projects: Callable[..., Iterable[Mapping]] | None = None,
) -> list[ThingsTodo]:
    """Every open to-do, with its breadcrumb resolved."""
    tasks = tasks or _default_tasks
    projects = projects or _default_projects

    area_of_project = {
        project["uuid"]: project.get("area_title") for project in projects()
    }
    # status=None: a heading marked complete/cancelled must still resolve the
    # breadcrumb for any of its to-dos that are themselves still open, or they
    # would silently misroute into the fallback list.
    heading_parents = {
        heading["uuid"]: (
            heading.get("project"),
            heading.get("project_title"),
            area_of_project.get(heading.get("project")),
        )
        for heading in tasks(type="heading", status=None)
    }

    rows = tasks(type="to-do", status="incomplete", include_items=True)

    return [_to_todo(row, heading_parents, area_of_project) for row in rows]


def _breadcrumb_fields(
    row: Mapping,
    heading_parents: Mapping[str, tuple[str | None, str | None, str | None]],
    area_of_project: Mapping[str, str | None],
) -> tuple[str | None, str | None, str | None, str | None]:
    """Resolve ``(area, project, heading, project_uuid)`` for one to-do.

    things.py sets ``heading_title`` *or* ``project_title`` on a to-do, never
    both, and never sets ``area_title`` on a to-do owned by a project. So the
    upper levels are looked up rather than read off the row.
    """
    heading_title = row.get("heading_title")
    if heading_title:
        project_uuid, project_title, area_title = heading_parents.get(
            row.get("heading"), (None, None, None)
        )
        return area_title, project_title, heading_title, project_uuid

    project_title = row.get("project_title")
    if project_title:
        project_uuid = row.get("project")
        return area_of_project.get(project_uuid), project_title, None, project_uuid

    return row.get("area_title"), None, None, None


def _outstanding_checklist(row: Mapping) -> tuple[str, ...]:
    """The checklist items still to do.

    Completed items are dropped rather than mirrored: they render as "☐ item",
    which would show finished work as outstanding.
    """
    return tuple(
        item["title"]
        for item in row.get("checklist") or ()
        if item.get("status") != "completed"
    )


def _to_todo(
    row: Mapping,
    heading_parents: Mapping[str, tuple[str | None, str | None, str | None]],
    area_of_project: Mapping[str, str | None],
) -> ThingsTodo:
    area_title, project_title, heading_title, project_uuid = _breadcrumb_fields(
        row, heading_parents, area_of_project
    )
    return ThingsTodo(
        uuid=row["uuid"],
        title=row.get("title") or "",
        notes=row.get("notes"),
        area_title=area_title,
        project_title=project_title,
        heading_title=heading_title,
        project_uuid=project_uuid,
        tags=tuple(row.get("tags") or ()),
        checklist=_outstanding_checklist(row),
        deadline=row.get("deadline"),
        start_date=row.get("start_date"),
    )

File: Sources/ThingsyncCore/ThingsSource.swift
/// Adapter: SQLite rows in, `ThingsTodo`/`ThingsProject` out.
///
/// The reader is injected so the whole transformation -- including
/// breadcrumb resolution -- is testable without a Things database. Unlike
/// the Python original, which delegated all Things schema knowledge to the
/// third-party things.py library, this port owns that knowledge outright
/// (see ThingsyncAdapters/ThingsDatabase.swift for the SQL half).

public enum ThingsTaskType: String, Sendable {
    case toDo = "to-do"
    case heading
    case project
}

public enum ThingsStatus: String, Sendable {
    case incomplete
    case completed
    case cancelled
}

public struct ThingsChecklistItemRow: Hashable, Sendable {
    public let title: String
    public let status: String

    public init(title: String, status: String) {
        self.title = title
        self.status = status
    }
}

/// One task row (a to-do or a heading), as read from the database. A
/// struct with optional fields, not an untyped dict, so the "heading_title
/// or project_title, never both" precedence logic in `breadcrumbFields`
/// stays a 1:1 port independent of SQLite.
public struct ThingsRow: Hashable, Sendable {
    public let uuid: String
    public let title: String
    public let notes: String?
    public let areaTitle: String?
    public let project: String?
    public let projectTitle: String?
    public let heading: String?
    public let headingTitle: String?
    public let tags: [String]
    public let checklist: [ThingsChecklistItemRow]
    public let deadline: String?
    public let startDate: String?

    public init(
        uuid: String,
        title: String,
        notes: String? = nil,
        areaTitle: String? = nil,
        project: String? = nil,
        projectTitle: String? = nil,
        heading: String? = nil,
        headingTitle: String? = nil,
        tags: [String] = [],
        checklist: [ThingsChecklistItemRow] = [],
        deadline: String? = nil,
        startDate: String? = nil
    ) {
        self.uuid = uuid
        self.title = title
        self.notes = notes
        self.areaTitle = areaTitle
        self.project = project
        self.projectTitle = projectTitle
        self.heading = heading
        self.headingTitle = headingTitle
        self.tags = tags
        self.checklist = checklist
        self.deadline = deadline
        self.startDate = startDate
    }
}

public struct ThingsProjectRow: Hashable, Sendable {
    public let uuid: String
    public let title: String?
    public let status: String?
    public let areaTitle: String?

    public init(uuid: String, title: String? = nil, status: String? = nil, areaTitle: String? = nil) {
        self.uuid = uuid
        self.title = title
        self.status = status
        self.areaTitle = areaTitle
    }
}

/// The seam things_source.py injected as bare `things.tasks`/`things.projects`
/// callables. The real implementation always resolves checklist items, so
/// unlike things.py there is no cheaper "don't bother with items" mode to
/// request -- `ThingsRow.checklist` is simply always populated.
public protocol ThingsReading {
    func projects(status: ThingsStatus?) throws -> [ThingsProjectRow]
    func tasks(type: ThingsTaskType, status: ThingsStatus?) throws -> [ThingsRow]
}

private struct HeadingParent {
    let projectUUID: String?
    let projectTitle: String?
    let areaTitle: String?
}

/// Resolve `(area, project, heading, project_uuid)` for one to-do.
///
/// Things sets `headingTitle` *or* `projectTitle` on a to-do, never both,
/// and never sets `areaTitle` on a to-do owned by a project. So the upper
/// levels are looked up rather than read off the row.
private func breadcrumbFields(
    row: ThingsRow,
    headingParents: [String: HeadingParent],
    areaOfProject: [String: String?]
) -> (areaTitle: String?, projectTitle: String?, headingTitle: String?, projectUUID: String?) {
    if let headingTitle = row.headingTitle, !headingTitle.isEmpty {
        let parent = row.heading.flatMap { headingParents[$0] }
        return (parent?.areaTitle, parent?.projectTitle, headingTitle, parent?.projectUUID)
    }

    if let projectTitle = row.projectTitle, !projectTitle.isEmpty {
        let projectUUID = row.project
        let areaTitle = projectUUID.flatMap { areaOfProject[$0] }.flatMap { $0 }
        return (areaTitle, projectTitle, nil, projectUUID)
    }

    return (row.areaTitle, nil, nil, nil)
}

/// The checklist items still to do.
///
/// Completed items are dropped rather than mirrored: they would render as
/// "☐ item", showing finished work as outstanding.
private func outstandingChecklist(row: ThingsRow) -> [String] {
    row.checklist.filter { $0.status != "completed" }.map(\.title)
}

private func toTodo(row: ThingsRow, headingParents: [String: HeadingParent], areaOfProject: [String: String?]) -> ThingsTodo {
    let fields = breadcrumbFields(row: row, headingParents: headingParents, areaOfProject: areaOfProject)
    return ThingsTodo(
        uuid: row.uuid,
        title: row.title,
        notes: row.notes,
        areaTitle: fields.areaTitle,
        projectTitle: fields.projectTitle,
        headingTitle: fields.headingTitle,
        projectUUID: fields.projectUUID,
        tags: row.tags,
        checklist: outstandingChecklist(row: row),
        deadline: row.deadline,
        startDate: row.startDate
    )
}

/// Every Things project, regardless of status.
///
/// A `nil` status filter is deliberate: things.py's own default of
/// `status="incomplete"` would silently drop completed and cancelled
/// projects -- exactly the ones whose Reminders list needs tearing down.
public func loadProjects(reader: ThingsReading) throws -> [ThingsProject] {
    try reader.projects(status: nil).map { row in
        ThingsProject(uuid: row.uuid, title: row.title ?? "", status: row.status ?? "incomplete")
    }
}

/// Every open to-do, with its breadcrumb resolved.
public func loadTodos(reader: ThingsReading) throws -> [ThingsTodo] {
    // Incomplete projects only: this is what makes areaOfProject only know
    // about incomplete projects' areas -- a to-do under a heading in a
    // completed project resolves areaTitle to nil. Preserved quirk, not a
    // bug to fix.
    var areaOfProject: [String: String?] = [:]
    for row in try reader.projects(status: .incomplete) {
        areaOfProject[row.uuid] = row.areaTitle
    }

    // status: nil -- a heading marked complete/cancelled must still resolve
    // the breadcrumb for any of its to-dos that are themselves still open,
    // or they would silently misroute into the fallback list.
    var headingParents: [String: HeadingParent] = [:]
    for heading in try reader.tasks(type: .heading, status: nil) {
        let areaTitle = heading.project.flatMap { areaOfProject[$0] }.flatMap { $0 }
        headingParents[heading.uuid] = HeadingParent(projectUUID: heading.project, projectTitle: heading.projectTitle, areaTitle: areaTitle)
    }

    let rows = try reader.tasks(type: .toDo, status: .incomplete)
    return rows.map { toTodo(row: $0, headingParents: headingParents, areaOfProject: areaOfProject) }
}
